// @serenity-js/protractor/lib/adapter/runner/TestRunnerDetector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunnerDetector = void 0;
/**
 * Detects the {@apilink TestRunnerAdapter} to use,
 * based on Protractor configuration.
 *
 * @group Integration
 */
class TestRunnerDetector {
    static protractorCliOptions() {
        return [
            TestRunnerDetector.cucumberOpts,
            TestRunnerDetector.jasmineNodeOpts,
            TestRunnerDetector.mochaOpts,
        ];
    }
    constructor(testRunnerLoader) {
        this.testRunnerLoader = testRunnerLoader;
    }
    runnerFor(config) {
        const specifiesRunnerFor = (type) => !!config.serenity &&
            !!config.serenity.runner &&
            config.serenity.runner === type;
        switch (true) {
            case specifiesRunnerFor('cucumber'):
                return this.useCucumber(config);
            case specifiesRunnerFor('jasmine'):
                return this.useJasmine(config);
            case specifiesRunnerFor('mocha'):
                return this.useMocha(config);
            case !!config.cucumberOpts:
                return this.useCucumber(config);
            case !!config.mochaOpts:
                return this.useMocha(config);
            case !!config.jasmineNodeOpts: // eslint-disable-line unicorn/no-useless-switch-case
            default:
                return this.useJasmine(config);
        }
    }
    useJasmine(config) {
        return this.testRunnerLoader.forJasmine(this.mergedConfigFor(config, TestRunnerDetector.jasmineNodeOpts));
    }
    useMocha(config) {
        return this.testRunnerLoader.forMocha(this.mergedConfigFor(config, TestRunnerDetector.mochaOpts));
    }
    useCucumber(config) {
        const serenityReportingServicesConfigured = config?.serenity?.crew?.length > 0;
        return this.testRunnerLoader.forCucumber(this.mergedConfigFor(config, TestRunnerDetector.cucumberOpts), {
            useStandardOutput: serenityReportingServicesConfigured,
            uniqueFormatterOutputs: this.multiCapabilitiesOrTestShardingEnabled(config),
        });
    }
    mergedConfigFor(config = {}, key) {
        return Object.assign({}, config[key], (config.capabilities || {})[key]);
    }
    multiCapabilitiesOrTestShardingEnabled(config) {
        return !!((Array.isArray(config.multiCapabilities) && config.multiCapabilities.length > 0)
            || typeof config.getMultiCapabilities === 'function'
            || config.capabilities?.shardTestFiles);
    }
}
exports.TestRunnerDetector = TestRunnerDetector;
TestRunnerDetector.cucumberOpts = 'cucumberOpts';
TestRunnerDetector.jasmineNodeOpts = 'jasmineNodeOpts';
TestRunnerDetector.mochaOpts = 'mochaOpts';
