// @serenity-js/web/lib/screenplay/models/PageElementsLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageElementsLocator = void 0;
const core_1 = require("@serenity-js/core");
const abilities_1 = require("../abilities");
/**
 * @group Models
 */
class PageElementsLocator extends core_1.Question {
    static fromDocumentRoot(selector) {
        return new PageElementsLocator(core_1.Question.about((0, core_1.d) `page elements located ${selector}`, async (actor) => {
            const bySelector = await actor.answer(selector);
            const currentPage = await abilities_1.BrowseTheWeb.as(actor).currentPage();
            return currentPage.locate(bySelector).locator;
        }));
    }
    constructor(locator) {
        super();
        this.locator = locator;
    }
    of(parent) {
        return new PageElementsLocator(core_1.Question.about(this.toString() + (0, core_1.d) ` of ${parent}`, async (actor) => {
            const locator = await actor.answer(this.locator);
            const parentElement = await actor.answer(parent);
            return locator.of(parentElement.locator);
        }));
    }
    async answeredBy(actor) {
        const resolved = await actor.answer(this.locator);
        return resolved.allElements();
    }
    describedAs(subject) {
        this.subject = subject;
        return this;
    }
    toString() {
        return this.subject ?? (0, core_1.d) `${this.locator}`;
    }
}
exports.PageElementsLocator = PageElementsLocator;
