// @serenity-js/core/lib/stage/crew/stream-reporter/StreamReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamReporter = void 0;
const tiny_types_1 = require("tiny-types");
const io_1 = require("../../../io");
/**
 * Serialises all the {@apilink DomainEvent} objects it receives and streams
 * them as [ndjson](http://ndjson.org/) to the output stream or file.
 *
 * Useful when debugging issues related to custom Serenity/JS test runner adapters.
 *
 * ## Registering Stream Reporter programmatically
 *
 * ```ts
 * import { configure, StreamReporter } from '@serenity-js/core'
 *
 * configure({
 *   crew: [
 *     new StreamReporter(process.stdout)
 *   ],
 * })
 * ```
 *
 * ## Writing Domain Events to a file
 *
 * ```ts
 * import { configure, StreamReporter } from '@serenity-js/core'
 * import fs = require('fs')
 *
 * configure({
 *   crew: [
 *     new StreamReporter(fs.createWriteStream('./events.ndjson'))
 *   ],
 * })
 * ```
 *
 * ## Using Stream Reporter with Playwright Test
 *
 * ```ts
 * // playwright.config.ts
 * import type { PlaywrightTestConfig } from '@serenity-js/playwright-test'
 *
 * const config: PlaywrightTestConfig = {
 *
 *   reporter: [
 *     [ '@serenity-js/playwright-test', {
 *       crew: [
 *         [ '@serenity-js/core:StreamReporter', { outputFile: './events.ndjson' } ]
 *       ]
 *       // other Serenity/JS config
 *     }]
 *   ],
 *   // other Playwright Test config
 * }
 * ```
 *
 * ## Using Stream Reporter with WebdriverIO
 *
 * ```ts
 * // wdio.conf.ts
 * import { WebdriverIOConfig } from '@serenity-js/webdriverio'
 *
 * export const config: WebdriverIOConfig = {
 *
 *   framework: '@serenity-js/webdriverio',
 *
 *   serenity: {
 *     crew: [
 *       '@serenity-js/serenity-bdd',
 *       [ '@serenity-js/core:StreamReporter', { outputFile: './events.ndjson' }]
 *     ]
 *     // other Serenity/JS config
 *   },
 *   // other WebdriverIO config
 * }
 * ```
 *
 * ## Using Stream Reporter with Protractor
 *
 * ```js
 * // protractor.conf.js
 * exports.config = {
 *   framework:     'custom',
 *   frameworkPath: require.resolve('@serenity-js/protractor/adapter'),
 *
 *   serenity: {
 *     crew: [
 *       [ '@serenity-js/core:StreamReporter', { outputFile: './events.ndjson' }]
 *     ],
 *     // other Serenity/JS config
 *   },
 *   // other Protractor config
 * }
 * ```
 *
 * @group Stage
 */
class StreamReporter {
    /**
     * Instantiates a `StreamReporter` outputting a stream of {@apilink DomainEvent|domain events}
     * to an `outputFile` at the given location.
     *
     * @param config
     */
    static fromJSON(config) {
        const outputFile = (0, tiny_types_1.ensure)('outputFile', config?.outputFile, (0, tiny_types_1.isDefined)(), (0, tiny_types_1.isString)());
        const cwd = config.cwd || process.cwd();
        const fs = new io_1.FileSystem(io_1.Path.from(cwd));
        return new StreamReporter(fs.createWriteStreamTo(io_1.Path.from(outputFile)));
    }
    /**
     * @param {stream~Writable} output
     *  A Writable stream that should receive the output
     *
     * @param {Stage} [stage]
     *  The stage this {@apilink StageCrewMember} should be assigned to
     */
    constructor(output = process.stdout, stage) {
        this.output = output;
        this.stage = stage;
    }
    /**
     * Creates a new instance of this {@apilink StageCrewMember} and assigns it to a given {@apilink Stage}.
     *
     * @param stage
     *  An instance of a {@apilink Stage} this {@apilink StageCrewMember} will be assigned to
     *
     * @returns {StageCrewMember}
     *  A new instance of this {@apilink StageCrewMember}
     */
    assignedTo(stage) {
        return new StreamReporter(this.output, stage);
    }
    /**
     * Handles {@apilink DomainEvent} objects emitted by the {@apilink StageManager}.
     *
     * @listens {DomainEvent}
     *
     * @param event
     */
    notifyOf(event) {
        this.output.write(JSON.stringify({ type: event.constructor.name, event: event.toJSON() }) + '\n');
    }
}
exports.StreamReporter = StreamReporter;
